src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp

```diff
@@ -0,0 +1,497 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines which types are allowed, whether operations which certain types are
+/// allowed, and other type-specific properties and configuration for
+/// `TensorExpression`s
+///
+/// \details
+/// To add support for a data type, modify the templates in this file and the
+/// arithmetic operator overloads as necessary. Then, add tests as appropriate.
+
+#pragma once
+
+#include <complex>
+#include <cstddef>
+#include <limits>
+#include <type_traits>
+
+#include "DataStructures/ComplexDataVector.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/VectorImpl.hpp"
+
+namespace tenex {
+template <typename DataType>
+struct NumberAsExpression;
+
+namespace detail {
+/// \brief Whether or not `TensorExpression`s supports using a given type as a
+/// numeric term
+///
+/// \details
+/// To make it possible to use a new numeric data type as a term in
+/// `TensorExpression`s, add the type to this alias and adjust other templates
+/// in this file, as necessary.
+///
+/// \tparam X the arithmetic data type
+template <typename X>
+using is_supported_number_datatype =
+    std::bool_constant<std::is_same_v<X, double> or
+                       std::is_same_v<X, std::complex<double>>>;
+
+/// \brief Whether or not `Tensor`s with the given data type are currently
+/// supported by `TensorExpression`s
+///
+/// \details
+/// To make it possible to use a new data type in a `Tensor` term in
+/// `TensorExpression`s, add the type to this alias and adjust other templates
+/// in this file, as necessary.
+///
+/// \tparam X the `Tensor` data type
+template <typename X>
+using is_supported_tensor_datatype = std::bool_constant<
+    std::is_same_v<X, double> or std::is_same_v<X, std::complex<double>> or
+    std::is_same_v<X, DataVector> or std::is_same_v<X, ComplexDataVector>>;
+
+/// \brief Whether or not the given type is a `VectorImpl` type
+///
+/// \tparam T the given type
+template <typename T>
+using is_vector = std::bool_constant<std::is_base_of_v<MarkAsVectorImpl, T>>;
+
+/// @{
+/// \brief If the given type is a derived `VectorImpl` type, get the base
+/// `VectorImpl` type, else return the given type
+///
+/// \tparam T the given type
+template <typename T, typename = void>
+struct upcast_if_derived_vector_type {
+  using type = T;
+};
+template <typename T>
+struct upcast_if_derived_vector_type<
+    T, typename std::enable_if<is_vector<T>::value>::type> {
```

Comment:
`git` warns to use `Requires` instead of `std::enable_if`, but I wasn't able to figure out how to get it to work in a template specialization. Let me know how I can use `Requires` here or suppress the warning!

Reply:
Thanks! I'll try this

Reply:
I also added a definition for the `upcast_if_derived_vector_type<T, false>` specialization when it's not a vector just to make it really clear what happens when

Reply:
Fine with me.